crates/but-github/src/token.rs

```diff
@@ -0,0 +1,75 @@
+use std::sync::Mutex;
+
+use anyhow::{Context, Result};
+use but_secret::{Sensitive, secret};
+use serde::{Deserialize, Serialize};
+
+/// Persist GitHub OAuth access tokens securely.
+pub fn persist_gh_access_token(login: &str, access_token: &Sensitive<String>) -> Result<()> {
+    let mut map = retrieve_github_oauth_access_token_map()?;
+    if let Some(entry) = map.iter_mut().find(|entry| entry.username == login) {
+        entry.access_token = access_token.clone().0;
+    } else {
+        map.push(UserToken {
+            username: login.to_string(),
+            access_token: access_token.clone().0,
+        });
+    }
+    persist_github_oauth_access_token_map(&map)
+}
+
+/// Delete a GitHub OAuth access token for a given username.
+pub fn delete_gh_access_token(login: &str) -> Result<()> {
+    let mut map = retrieve_github_oauth_access_token_map()?;
+    map.retain(|entry| entry.username != login);
+    persist_github_oauth_access_token_map(&map)
+}
+
+/// Retrieve a GitHub OAuth access token for a given username.
+pub fn get_gh_access_token(login: &str) -> Result<Option<Sensitive<String>>> {
+    let map = retrieve_github_oauth_access_token_map()?;
+    Ok(map
+        .into_iter()
+        .find(|entry| entry.username == login)
+        .map(|entry| Sensitive(entry.access_token)))
+}
+
+pub fn list_known_github_usernames() -> Result<Vec<String>> {
+    let map = retrieve_github_oauth_access_token_map()?;
+    Ok(map.into_iter().map(|entry| entry.username).collect())
+}
+
+const GITHUB_OAUTH_ACCESS_TOKEN_MAP_KEY: &str = "github_oauth_access_token_map";
+
+#[derive(Debug, Deserialize, Serialize, Clone)]
+struct UserToken {
+    pub username: String,
+    pub access_token: String,
+}
```

Comment:
Can we differentiate the token type by looking at the `access_token`? I am asking because of the semantic difference between device tokens and personal access tokens.